diagnostic_aggregator/src/aggregator.cpp

```diff
@@ -107,13 +106,77 @@ void Aggregator::diagCallback(const diagnostic_msgs::DiagnosticArray::ConstPtr&
   }
 }
 
-Aggregator::~Aggregator() 
+Aggregator::~Aggregator()
 {
   if (analyzer_group_) delete analyzer_group_;
 
   if (other_analyzer_) delete other_analyzer_;
 }
 
+
+void Aggregator::bondBroken()
+{
+  boost::mutex::scoped_lock lock(mutex_); // Possibility of multiple bonds breaking at once
+
+  vector<boost::shared_ptr<bond::Bond> >::iterator it;
+  for (it = bonds_.begin(); it != bonds_.end();)
+  {
+    if ((*it)->isBroken())
+    {
+      string id = (*it)->getId();
+      ROS_DEBUG("bond for namespace %s was broken", id.c_str());
+      analyzer_group_->removeAnalyzer(added_analyzers_.find(id)->second);
+      added_analyzers_.erase(id);
+      bonds_.erase(it); // it points to next in list (or end())
+    } else {
+      // increment in afterthought will go past bonds_.end() if erase()ing last
+      // element, so do it here instead
+      ++it;
+    }
+  }
+}
+
+void Aggregator::bondFormed(){}
+
+bool Aggregator::addDiagnostics(diagnostic_msgs::AddDiagnostics::Request &req,
+				diagnostic_msgs::AddDiagnostics::Response &res)
+{
+  // Wait for connections on this bond. Wait until it's formed to put it in the
+  // bond vector
```

Comment:
The comment is slightly out of date from some previous attempts at implementing the functionality. I wanted to ensure that the bond was successfully formed in the service call, so that the information could be passed back to the caller. Using the `waitUntilFormed` function in the service call doesn't work, since once inside it there is no way to receive messages on the bond topic unless there is another thread which is calling `ros::spin()` periodically, so I abandoned that approach.

Doing the initialisation in `bondFormed` means that the service call can only provide minimal information on return; probably just whether an attempt was made to add parameters from a namespace which was used before. If the `AnalyzerGroup` fails to initialise, the only way that information can be passed back to the caller is by breaking the bond, which I don't think is ideal.